fastcrypto/src/secp256r1.rs

```diff
@@ -0,0 +1,495 @@
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(tag = "type")] // necessary so as not to deser under a != type
+pub struct Secp256r1KeyPair {
+    pub name: Secp256r1PublicKey,
+    pub secret: Secp256r1PrivateKey,
+}
+
+impl EncodeDecodeBase64 for Secp256r1KeyPair {
+    fn encode_base64(&self) -> String {
+        let mut bytes: Vec<u8> = Vec::new();
+        bytes.extend_from_slice(self.secret.as_ref());
+        bytes.extend_from_slice(self.name.as_ref());
+        Base64::encode(&bytes[..])
+    }
+
+    fn decode_base64(value: &str) -> Result<Self, eyre::Report> {
+        keypair_decode_base64(value)
+    }
+}
+
+impl KeyPair for Secp256r1KeyPair {
+    type PubKey = Secp256r1PublicKey;
+    type PrivKey = Secp256r1PrivateKey;
+    type Sig = Secp256r1Signature;
+
+    fn public(&'_ self) -> &'_ Self::PubKey {
+        &self.name
+    }
+
+    fn private(self) -> Self::PrivKey {
+        Secp256r1PrivateKey::from_bytes(self.secret.as_ref()).unwrap()
+    }
+
+    #[cfg(feature = "copy_key")]
+    fn copy(&self) -> Self {
+        Secp256r1KeyPair {
+            name: self.name.clone(),
+            secret: Secp256r1PrivateKey::from_bytes(self.secret.as_ref()).unwrap(),
+        }
+    }
+
+    fn generate<R: AllowedRng>(rng: &mut R) -> Self {
+        let privkey = ExternalSecretKey::random(rng);
+        let pubkey = ExternalPublicKey::from(&privkey);
+
+        Secp256r1KeyPair {
+            name: Secp256r1PublicKey {
+                pubkey,
+                bytes: OnceCell::new(),
+            },
+            secret: Secp256r1PrivateKey {
+                privkey,
+                bytes: OnceCell::new(),
+            },
+        }
+    }
+}
+
+impl FromStr for Secp256r1KeyPair {
+    type Err = eyre::Report;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let kp = Self::decode_base64(s).map_err(|e| eyre::eyre!("{}", e.to_string()))?;
+        Ok(kp)
+    }
+}
+
+impl Signer<Secp256r1Signature> for Secp256r1KeyPair {
+    fn try_sign(&self, msg: &[u8]) -> Result<Secp256r1Signature, signature::Error> {
+        Ok(Secp256r1Signature {
+            sig: self.secret.privkey.sign(msg),
+            bytes: OnceCell::new(),
+        })
+    }
+}
+
+impl TryFrom<Secp256r1PublicKeyBytes> for Secp256r1PublicKey {
+    type Error = signature::Error;
+
+    fn try_from(bytes: Secp256r1PublicKeyBytes) -> Result<Secp256r1PublicKey, Self::Error> {
+        Secp256r1PublicKey::from_bytes(bytes.as_ref()).map_err(|_| Self::Error::new())
+    }
+}
+
+impl From<&Secp256r1PublicKey> for Secp256r1PublicKeyBytes {
+    fn from(pk: &Secp256r1PublicKey) -> Self {
+        Secp256r1PublicKeyBytes::from_bytes(pk.as_ref()).unwrap()
+    }
+}
+
+impl From<Secp256r1PrivateKey> for Secp256r1KeyPair {
+    fn from(secret: Secp256r1PrivateKey) -> Self {
+        let name = Secp256r1PublicKey::from(&secret);
+        Secp256r1KeyPair { name, secret }
+    }
+}
+
+impl Secp256r1Signature {
+    /// Recover public key from signature
```

Comment:
Can you refer to the original impl this was based, just to ensure no malleability and consistency with the secpK1 lib. Ideally we should copy paste that logic (if not).
@joyqvq please have a look as well

Reply:
I will. It is mostly based on the specs (section 4.1.6 in https://www.secg.org/sec1-v2.pdf) and also on an implementation in the k256 crate. I have two somewhat related questions regarding this @kchalkias :
* I'm assuming we wan't to recover and check public keys every time because of the attack you found recently, but both verifying AND recovering public keys seem to be at least double work. Should we make public key recovery optional?
* Should we add a recovery id byte to signatures as it is done in secp256k1? Since the endgoal is interoperability with hardware wallets, I guess the answer to this depends on how the hardware wallets are handling this, but I haven't been able to find some good documentation.